### PR TITLE
feat: add required vars support

### LINKED ISSUE #29: Support variable requirements in referenced blocks
## Background

The reference-able `content` blocks, defined on the root level of the codebase, need to access specific data points from the context. The content block might not know the exact keys available in the context, making the logic inside the block fragile and error-prone. This makes block reuse difficult.

The blocks that allow `vars` blocks (`content`, `document`, `section` blocks) and that can be referenced, should be able to declare the requirements: a list of variables a block expects to find in the context during execution.

## Design

We introduce a new arrtibute to the `content`, `document` and `section` blocks:

`required_vars` -- (optional) a list of strings

`required_vars` defines the names of the variables that the block expectes to find in the context under `.vars` namespace.

For example:

```hcl
content text "hello" {
  text = "Hello, {{ .vars.name }}"
  required_vars = ["name"]
}

content text "greeting" {
  text = "Greetings, {{ .vars.other_name }}"
  required_vars = ["other_name"]
}

document "bar" {
  vars {
    name = "Bruce"
  }

  content ref {
    base = content.text.hello
  }

  content ref {
    vars {
      other_name = query_jq(".vars.name")
    }
    base = content.text.greetings
  }
}
```
renders into
```
Hello, Bruce
Greetings, Bruce
```

while 
```hcl
content text "hello" {
  text = "Hello, {{ .vars.name }}"
  required_vars = ["name"]
}

document "bar" {
  content ref {
    base = content.text.hello
  }
}
```
throws an error saying that `content.text.hello` block requires `name` var which is not set.

`required_vars` is optional -- if not set in the block, the block behaves as it does currently 

Note: Required variables should be asserted after all vars / attributes in `ref` block are resolved.

## References
- https://github.com/blackstork-io/fabric/issues/166

### PR DESCRIPTION
**Pull request by anasmuhmd**
closes #29

### CODE REVIEW

File: engine/engine_vars_test.go [hunk 1]
```diff
@@ -371,4 +371,44 @@ func TestEngineVarsHandling(t *testing.T) {
...
 		diagtest.Asserts{},
 	)
+	renderTest(
+		t, "required vars in document",
+		[]string{`
+			document "example" {
+				vars {
+					now = "2021-01-01"
+				}
+				required_vars = ["now"]
```
Comment (traut) [inline engine/engine_vars_test.go:381]: Interesting, this might be one of the edge cases -- even though documents can be referenced with `ref` keyword, the resolution of variables might not be obvious. I think it's good to have a test for it!

Comment (anasmuhmd) [reply engine/engine_vars_test.go:381]: Makes sense, I started with document because the issue mentioned it. I'll work on content and section blocks

**Comment by traut** [comment]
@Andrew-Morozko any feedback on the approach here, before Anas dives deeper?

File: eval/document.go [hunk 2]
```diff
@@ -52,7 +53,7 @@ func (doc *Document) FetchData(ctx context.Context) (plugindata.Data, diagnostic
 	return result, diags
 }
 
-func (doc *Document) RenderContent(ctx context.Context, docDataCtx plugindata.Map) (plugin.Content, plugindata.Data, diagnostics.Diag) {
+func (doc *Document) RenderContent(ctx context.Context, docDataCtx plugindata.Map, name string) (plugin.Content, plugindata.Data, diagnostics.Diag) {
```
Comment (Andrew-Morozko) [inline eval/document.go:56]: Minor nitpick: I don't think that adding a `name` argument here is worth it just to display it in a diagnostic. 
I suggest adding the ability to refer back to the document block through our structures:

* add a `Source *definitions.ParsedDocument` to the `Document` struct [here](https://github.com/blackstork-io/fabric/blob/1aceed03feec09474928d932d1fdbe976dcfe58d/eval/document.go#L17), set it to `node` [here](https://github.com/blackstork-io/fabric/blob/1aceed03feec09474928d932d1fdbe976dcfe58d/eval/document.go#L138)
* add a `Raw *definitions.Document` [here](https://github.com/blackstork-io/fabric/blob/1aceed03feec09474928d932d1fdbe976dcfe58d/parser/definitions/parsed_document.go#L3), set it to `d` [here](https://github.com/blackstork-io/fabric/blob/1aceed03feec09474928d932d1fdbe976dcfe58d/parser/parse_document.go#L16)

File: eval/document.go [hunk 3]
```diff
@@ -74,6 +75,21 @@ func (doc *Document) RenderContent(ctx context.Context, docDataCtx plugindata.Ma
...
+	// verify required vars
+	if len(doc.RequiredVars) > 0 {
+		vars, varsPresent := docDataCtx["vars"].(plugindata.Map)
+		for _, reqVar := range doc.RequiredVars {
+			if !varsPresent || vars[reqVar] == nil {
+				diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Missing required variable",
+					Detail:   "document." + name + " block requires '" + reqVar + "' var which is not set.",
+					Subject:  nil,
```
Comment (Andrew-Morozko) [inline eval/document.go:87]: And here you would be able to use `Subject: doc.Source.Raw.Block.Range()` to point the user at the offending block

**Review by Andrew-Morozko — COMMENTED** [review]
LGTM, nicely done!

**Review by Andrew-Morozko — APPROVED** [review]
LGTM! Thanks for your contribution.

There are some problems with linter in CI that are fixed in #231. It will land as soon as @traut got a free minute, after that I ask you to rebase your changes on top of that and we will merge this PR in.

**Comment by traut** [comment]
#231 is merged!
